src/db.c

```diff
@@ -1048,15 +1125,45 @@ void scanGenericCommand(client *c, robj *o, unsigned long cursor) {
     listRelease(keys);
 }
 
+void addSlotIdToCursor(int slot, unsigned long long int *cursor) {
+    /* Slot id can be -1 if there are no more slots to visit. */
+    if (slot >= 0) {
+        *cursor = (*cursor << CLUSTER_SLOT_MASK_BITS) | slot;
+    }
+}
+
+int getAndClearSlotIdFromCursor(unsigned long long int *cursor) {
+    int slot = (int) (*cursor & CLUSTER_SLOT_MASK);
+    *cursor = ((*cursor) & ~CLUSTER_SLOT_MASK) >> CLUSTER_SLOT_MASK_BITS;
+    return slot;
+}
```

Comment:
i'm probably missing something.
it looks to me as if the first function shifts the cursor only in cluster mode, and leave it low in non-cluster mode.
but in the other function it looks like we always shift.

if it's not already like it, i suggest that for non-cluster scan, we'll leave the cursor unaffected.

Reply:
Cursor in non-cluster mode should not be affected at all. As you said, first function will never touch it, and the second one will always return 0 when cluster mode is disabled.  We can add defensive code that would check for `cluster_enabled` status explicitly, but I think it would be an overkill.

Reply:
To be clear, in non-cluster mode, the cursor is shifted by 14 bits too and slot is encoded in the cursor, but slot is always 0.

(During scan, dbGetNextUnvisitedSlot returns the next dict and slot, which is never more than 0 in non-cluster mode since it uses `db->dict_count`. Btw, it's not necessary to store `dict_count` in the db struct. It is always `server.cluster_enabled ? CLUSTER_SLOTS : 1` and `cluster_enabled` cannot be changed at runtime.)

Reply:
aren't you contradicting each other?
anyway, i didn't bother to look too deeply as to when `int slot` would be negative, but the first one seems to only shift the cursor when slot is non-negative, and the second one always shifts it.
so either something is wrong here, or maybe some comments can clear it up.
in any case, i think that we should aim for the non-cluster mode to have unshifted cursors.

Reply:
Yes, I'm contradicting. :D I shouldn't have said "to be clear", but sometging else. I understand it differently but @vitarb please correct me if I'm wrong.

Slot is -1 when scan is finished. I believe cursor is always zero in this case. I think we should add an assert here.

Btw, I think dbGetNextUnvisitedSlot could set slot to 0 instead of -1 when there are no more slots. I don't see why we need the -1 value.

Reply:
I see there is some confusion. Let me take another look at this code and try to make it a bit more readable and easy to understand.

Reply:
for the record, i still don't get it.
getAndClearSlotIdFromCursor always shifts the cursor by 14 bits, even in non-cluster mode.